Trigger change() on attribute_priceTE when clicking save button to handle broken localization packs.

Prestaworks opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Fix for issue https://github.com/PrestaShop/PrestaShop/issues/23120 where localization settings breaks product page in BO.
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #23120.
| How to test?      | See issue for details.
| Possible impacts? | product page and management of combinations in BO.


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/23137)
<!-- Reviewable:end -->

Prestaworks committed:
Update form.js

PierreRambaud reviewed (changes requested):
Hi and thanks for your contribution.
Could you please use a proper title and description for a better understanding? :)

Prestaworks commented:
> 
> 
> Hi and thanks for your contribution.
> Could you please use a proper title and description for a better understanding? :)

better?

matks commented:
@Prestaworks I think @PierreRambaud was rather looking for a title that explains what happens, like "Trigger change() on attribute_priceTE in product page to handle swedish combination bug"

When we build a release, the Pull Request titles are used to build a Changelog. People use the Changelog to understand what changed between two versions, so we need the titles to be explanatory 😉  .

Prestaworks commented:
> 
> 
> @Prestaworks I think @PierreRambaud was rather looking for a title that explains what happens, like "Trigger change() on attribute_priceTE in product page to handle swedish combination bug"
> 
> When we build a release, the Pull Request titles are used to build a Changelog. People use the Changelog to understand what changed between two versions, so we need the titles to be explanatory 😉 .

I understand :) I hope this title will be helpfull.

PierreRambaud reviewed (dismissed):
Thank you :)

khouloudbelguith commented:
Hi @Prestaworks,

This issue is not fixed yet.
I followed these steps:

1. Rebuild assets
2. Install the PR with English language & France country 
3. Go to BO > International > Localization page > Upload Sweden pack
4. Go to BO > Your profile > change the language of the employee to Swedish 
5. Go to BO > Catalog > Edit the demo product `demo_1`
6. Change the price to a negative price change on one of the combinations
7. Save => ok
8. Click on `Gå till katalog` => redirection to the catalog list page
9. Edit the product > Edit the last combination updated
10. See error => the impact tax included is empty => NOK
11. See an error 500 in the console 

![image](https://user-images.githubusercontent.com/16067358/107505778-7f9d8800-6b9d-11eb-8fb9-a59c42956d78.png)

I attached a screen record
https://drive.google.com/file/d/1JNwSaCsj-N1qFN8dA746M7MIGlnLUrCj/view?usp=sharing

Also, I have an issue with the Swedish language reproduced **with** or **without** PR following these steps:
1. Rebuild assets
2. Install the PR with English language & France country 
3. Go to BO > International > Localization page > Upload Sweden pack
4. Go to BO > Your profile > change the language of the employee to Swedish 
5. Go to BO > Catalog > Add a new product
6. Add a specific price with impact on price = -2 
![image](https://user-images.githubusercontent.com/16067358/107506651-ac9e6a80-6b9e-11eb-8287-a0716259b8ff.png)
7. Update this specific price by applying a discount of -2 euro 
![image](https://user-images.githubusercontent.com/16067358/107506738-d35ca100-6b9e-11eb-92b1-9ae765c3fb00.png)
8. Save => see error 500 in the console 
![image](https://user-images.githubusercontent.com/16067358/107506773-e1122680-6b9e-11eb-9cca-57edf2d7c581.png)

If the English language is selected => we have a clear alert & an error 400 in the console 
![image](https://user-images.githubusercontent.com/16067358/107506925-1ae32d00-6b9f-11eb-93fb-22020c2f9865.png)


Thank you!

Prestaworks committed:
complement to issue 23120

This will also fix the issue reported by khouloudbelguith regarding the edit of combinations.

Prestaworks commented:
Hi @khouloudbelguith 
Good to see that you have your eyes open :) I have added a change to one more file to adress the issue with the edit of combinations.

However, the issue with negative specific price, I don't think that's supposed to be possible? as typing a number there turns into a discount, so a negative value would be an price increase and not a discount.

khouloudbelguith commented:
@Prestaworks,

>However, the issue with negative specific price, I don't think that's supposed to be possible? as typing a number there turns into a discount, so a negative value would be an price increase and not a discount.

I think we can create a new issue about this problem and check the Product team about the exact behavior.
What do you think?

Thanks!

Prestaworks commented:
> 
> 
> @Prestaworks,
> 
> > However, the issue with negative specific price, I don't think that's supposed to be possible? as typing a number there turns into a discount, so a negative value would be an price increase and not a discount.
> 
> I think we can create a new issue about this problem and check the Product team about the exact behavior.
> What do you think?
> 
> Thanks!

Yes, that's probobly a good idea.

khouloudbelguith commented:
Hi @Prestaworks,

I have some issues:

1. Rebuild assets
2. Install the PR with English language & France country
3. Go to BO > International > Localization page > Upload Sweden pack
4. Edit a product `demo_3` > Change the price to a negative price change on one of the combinations
5. Save => ok
6. Go to BO > Your profile > change the language of the employee to Swedish
7. Go to BO > Catalog > Edit the previous product
8. Edit the last combination updated
9. See error => the impact tax included is empty => NOK
10. See an error 500 in the console

I attached a screen record
https://drive.google.com/file/d/1iyEK5z638JykqpEPAvxciusJtr66PELg/view?usp=sharing

PS: when I install the shop with Swedish language => this issue is not reproduced

Also, in this PR, `Impact on price per unit (tax excl.)`, `Impact on weight` should be fixed?

Steps to reproduce the issue:

1. Rebuild assets
2. Install the PR with English language & France country
3. Go to BO > International > Localization page > Upload Sweden pack
4. Go to BO > Your profile > change the language of the employee to Swedish
5. Go to BO > Catalog > Edit the demo product `demo_1`
6. Change the Impact on price per unit (tax excl.)` & the `Impact on weight` to a negative impact change on one of the combinations
7. Save => ok
8. Click on `Gå till katalog` => redirection to the catalog list page
9. Edit the product
10. Save 
11. See error

![image](https://user-images.githubusercontent.com/16067358/107661400-8b667880-6c89-11eb-85ab-16f7bb927811.png)

PS: when English is selected => ok

Thank you!

Prestaworks commented:
Hi @khouloudbelguith 
I can see that you get some kind of JS error in the movie, can you see what error it is in the console and if it is related to this?
It seems almost like you found another issue when changing language on the user.

I will see what I can do about the other fields you mention also.

khouloudbelguith commented:
Hi @Prestaworks,

Thanks for your feedback.
>I can see that you get some kind of JS error in the movie, can you see what error it is in the console and if it is related to this?
It seems almost like you found another issue when changing language on the user.

In fact, I just tried another install, but the first issue is not reproduced anymore
I'm sorry, but it could be a cache issue or a problem in uploading the Swedish pack.
I attached a screen record https://drive.google.com/file/d/1-A9su8CuOvWu8wQF_La0PNmr3uKYH5_k/view?usp=sharing

>I will see what I can do about the other fields you mention also.

This issue is still exists.

Thank you!

Prestaworks committed:
Added more fields

Made changes to fix bad - sign for more fields. also added support for negative wholesale price impact.

Prestaworks commented:
I added support now also for negative price impact on the wholesale_price, the DB allows it, but the class was not.
All the fields should also now have a fix on the save click for the bad - char, including weight impact.

atomiix commented:
Closed and reopened to relaunch tests

atomiix commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -838,6 +838,9 @@ var form = (function() {
       });
 
       elem.submit(function(event) {
+        $.each($('input.attribute_wholesale_price, input.attribute_priceTE, input.attribute_priceTI, input.attribute_unity, input.attribute_weight'), function () {
+          $(this).val($(this).val().replace("−", "-"));
+        });
```
What do you think about extracting that part into a function that you could call from there and from line `860`? It would be more DRY :)

Prestaworks replied:
Good idea.

Prestaworks committed:
Update form.js

khouloudbelguith commented:
Hi @Prestaworks,

I have always some issues:
I tried to install two shops:
1. Shop1: English language & France country
2. Shop2 Swedish language & France country

Steps to reproduce the issue:

1. In the first shop, `shop1` > I uploaded the Sweden pack in the BO > International > Localization page
2. Change the employee language to `Swedish`
3. Go to BO > Catalog > Products > Create new product with combinations
4. Set the price of the original product
5. Change the price to a negative price change on all combinations
6. Click on save 
7. Click on `Gå till katalog` => redirection to the catalog list page
8. Edit the product > Edit the last combination updated
9. See error => the impact tax included is empty => NOK
10. Save
11. See an error 500 in the console

Shop1: https://drive.google.com/file/d/1idZrvr4zdnvoSof5WnTlHIUa-0E24h7D/view?usp=sharing


Shop2: https://drive.google.com/file/d/17NNcyodP8ciOhIo1G1ppTOxCObImHtnV/view?usp=sharing

1. Go to BO > Catalog > Products > Create new product with combinations
2. Set the price of the original product
3. Change the price to a negative price change on all combinations
4. Click on save 
5. Click on `Gå till katalog` => redirection to the catalog list page
6. Edit the product > Edit the last combination updated
7. See error => the impact tax included is empty => NOK
8. Save => ok

Thank you!

Prestaworks commented:
@khouloudbelguith is that the same issue you had earlier? that solved itself in another installation?

khouloudbelguith commented:
Hi @Prestaworks,

>@khouloudbelguith is that the same issue you had earlier? that solved itself in another installation?

Yes, but the issue `the impact tax included is empty` is reproduced in two shops, the first one installed with English language & the second one installed with the Swedish language.

Thank you!

Prestaworks committed:
Adding run of fix on load of iframe

Prestaworks commented:
@khouloudbelguith I have added a fix for that issue now as well.

atomiix commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -791,7 +791,7 @@ var form = (function() {
     }
   }
-
+  
```
```suggestion

```

atomiix commented on admin-dev/themes/default/js/bundle/product/product-combinations.js:
```diff
@@ -1,3 +1,11 @@
+ */
+function replaceBadLocaleCharacters() {
+    $.each($('input.attribute_wholesale_price, input.attribute_priceTE, input.attribute_priceTI, input.attribute_unity, input.attribute_weight'), function () {
+        $(this).val($(this).val().replace("−", "-"));
```
```suggestion
        $(this).val($(this).val().replace("−", "-")); // replace U+002D with U+2212
```

atomiix reviewed (commented):
Just added a comment as a suggestion to make it clearer

Prestaworks committed:
Update product-combinations.js

atomiix committed:
Remove empty space

khouloudbelguith commented:
Hi @Prestaworks,

It is ok :heavy_check_mark: 
I tried with two shops:

1. Shop1: English language & France country
2. Shop2 Swedish language & France country

https://drive.google.com/file/d/1m0MY2oWJUChZeSyj1dw3gcSUDpnauVW_/view?usp=sharing

Thank you!

Progi1984 commented:
Thanks @Prestaworks & @khouloudbelguith
